File: 338-counting-bits/338-counting-bits.js
/**
 * @param {number} n
 * @return {number[]}
 */
var countBits = function(n) {
    if (n === 0) return [0];
    const ans = new Array(n+1).fill(0);
    let offset = 1;
    
    for (let i = 1; i <= n; i++) {
        if (i === offset * 2) {
            offset = i;
        }
        ans[i] = 1 + ans[i - offset];
    }
    
    return ans;
};

File: 338-counting-bits/338-counting-bits.rb
# @param {Integer} n
# @return {Integer[]}
def count_bits(n)
    return [0] if n == 0
    ans = Array.new(n+1, 0)
    offset = 1
    (1..n).each do |i|
        if i == offset * 2
            offset = i
        end
        ans[i] = 1 + ans[i - offset]
    end
    return ans
end